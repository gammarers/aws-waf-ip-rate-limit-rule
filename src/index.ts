import * as wafv2 from 'aws-cdk-lib/aws-wafv2';

export interface RuleConfig {
  readonly priority: number;
  readonly ruleName?: string;
  readonly cloudWatchMetricsName?: string;
}

export interface WAFIPRateLimitRuleProps {
  readonly rateLimit: number;
}

export class WAFIPRateLimitRule {

  constructor(private props: WAFIPRateLimitRuleProps) {
  }

  blockRule(config: RuleConfig): wafv2.CfnWebACL.RuleProperty {
    return {
      name: config.ruleName || 'block-rate-limit-rule',
      priority: config.priority,
      action: {
        block: {
          //  customResponse: {
          //    CustomResponseBodyKey: 'ip-restrict',
          //    ResponseCode: 403,
          //  },
        },
      },
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        metricName: config.cloudWatchMetricsName || 'BlockRateLimitRuleMetric',
        sampledRequestsEnabled: true,
      },
      statement: {
        rateBasedStatement: {
          aggregateKeyType: 'IP',
          limit: this.props.rateLimit,
        },
      },
    };
  }
}

//export class WafIpRateLimitRuleGroup extends waf.CfnRuleGroup {
//  constructor(scope: Construct, id: string, props: WafIpRateLimitRuleGroupProps) {
//    super(scope, id, {
//      name: props?.name,
//      description: 'rate limit rule group',
//      scope: ((): string => {
//        switch (props.scope) {
//          case Scope.GLOBAL:
//            return 'CLOUDFRONT';
//          case Scope.REGIONAL:
//            return 'REGIONAL';
//        }
//      })(),
//      capacity: 10,
//      customResponseBodies: {
//        ['ip-restrict']: {
//          contentType: 'TEXT_PLAIN',
//          content: 'Sorry, You Are Not Allowed to Access This Service.',
//        },
//      },
//      rules: [
//        {
//          priority: 10,
//          name: 'rate-limit-rule',
//          action: {
//            block: {
//              CustomResponse: {
//                CustomResponseBodyKey: 'ip-restrict',
//                ResponseCode: 403,
//              },
//            },
//          },
//          visibilityConfig: {
//            cloudWatchMetricsEnabled: true,
//            sampledRequestsEnabled: true,
//            metricName: 'WafRateLimitRule',
//          },
//          statement: {
//            rateBasedStatement: {
//              aggregateKeyType: 'IP',
//              limit: props?.rateLimitCount ?? 1000,
//            },
//          },
//        },
//      ],
//      visibilityConfig: {
//        cloudWatchMetricsEnabled: true,
//        sampledRequestsEnabled: true,
//        metricName: 'RateLimitRule',
//      },
//      tags: [],
//    });
//  }
//}