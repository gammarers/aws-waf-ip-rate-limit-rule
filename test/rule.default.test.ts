import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as wafv2 from 'aws-cdk-lib/aws-wafv2';
import { WAFIPRateLimitRule } from '../src';

describe('Web Acl rule default group testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack');

  const ipRateLimitRule = new WAFIPRateLimitRule({
    rateLimit: 100,
  });

  new wafv2.CfnWebACL(stack, 'WebACL', {
    defaultAction: { allow: {} },
    scope: 'CLOUD_FRONT',
    name: 'WebAclWithCustomRules',
    visibilityConfig: {
      cloudWatchMetricsEnabled: true,
      metricName: 'WebAclMetric',
      sampledRequestsEnabled: true,
    },
    rules: [
      ipRateLimitRule.blockRule({
        priority: 1,
      }),
    ],
  });

  const template = Template.fromStack(stack);

  it('Should have WAF WebACL Rule', async () => {
    template.hasResourceProperties('AWS::WAFv2::WebACL', {
      Rules: [
        {
          Action: {
            Block: {},
          },
          Name: 'block-rate-limit-rule',
          Priority: 1,
          Statement: {
            RateBasedStatement: {
              AggregateKeyType: 'IP',
              Limit: 100,
            },
          },
          VisibilityConfig: {
            CloudWatchMetricsEnabled: true,
            MetricName: 'BlockRateLimitRuleMetric',
            SampledRequestsEnabled: true,
          },
        },
      ],
    });
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});