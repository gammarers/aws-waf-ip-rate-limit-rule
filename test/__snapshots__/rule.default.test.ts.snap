// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web Acl rule default group testing Should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "WebACL": {
      "Properties": {
        "DefaultAction": {
          "Allow": {},
        },
        "Name": "WebAclWithCustomRules",
        "Rules": [
          {
            "Action": {
              "Block": {},
            },
            "Name": "block-rate-limit-rule",
            "Priority": 1,
            "Statement": {
              "RateBasedStatement": {
                "AggregateKeyType": "IP",
                "Limit": 100,
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "BlockRateLimitRuleMetric",
              "SampledRequestsEnabled": true,
            },
          },
        ],
        "Scope": "CLOUD_FRONT",
        "VisibilityConfig": {
          "CloudWatchMetricsEnabled": true,
          "MetricName": "WebAclMetric",
          "SampledRequestsEnabled": true,
        },
      },
      "Type": "AWS::WAFv2::WebACL",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
